File: vccpython/components/expansionvalve.py
from .port import *


class ExpansionValve:

    energy = "None"
    devtype = "EXPANSIONVALVE"

    def __init__(self, dictDev):
        """ Initializes the ExpansionValve """
        self.name = dictDev['name']
        self.iPort = [Port(dictDev['iPort'])]
        self.oPort = [Port(dictDev['oPort'])]
        # map the port's name(str) to the obj
        self.portdict = {
            "iPort": self.iPort,
            "oPort": self.oPort
        }

    def state(self):
        self.oPort[0].h = self.iPort[0].h

    def balance(self):
        """ mass and energy balance  """
        if self.iPort[0].mdot is not None:
            self.oPort[0].mdot = self.iPort[0].mdot
        elif self.oPort[0].mdot is not None:
            self.iPort[0].mdot = self.oPort[0].mdot

    def __str__(self):
        result = '\n' + self.name
        result += '\n' +" PORTS "+Port.title
        result += '\n' +" iPort "+ self.iPort[0].__str__()
        result += '\n' +" oPort "+ self.oPort[0].__str__()
        return result

File: vcccpp/src/expansionvalve.cpp
/*
  expansionvalve.cpp

*/

#include "expansionvalve.hpp"

ExpansionValve::ExpansionValve(umComponent dictComp)
{
    name = any_cast<const char *>(dictComp["name"]);
    iPort = new Port(any_cast<mPort>(dictComp["iPort"]));
    oPort = new Port(any_cast<mPort>(dictComp["oPort"]));
    portdict = {{"iPort", iPort},
                {"oPort", oPort}};
    
 
}

ExpansionValve::~ExpansionValve()
{
    delete iPort;
    delete oPort;
}

void ExpansionValve::state()
{
    oPort->h = iPort->h;
}

void ExpansionValve::balance()
{
    // mass and energy balance
    // mass balance
    if (!isnan(iPort->mdot))
    {
        oPort->mdot = iPort->mdot;
    }
    else
    {
        if (!isnan(oPort->mdot))
            iPort->mdot = oPort->mdot;
    }
}

void ExpansionValve::setportaddress()
{
  if (iPort!=portdict["iPort"])
      iPort=portdict["iPort"];
  if (oPort!=portdict["oPort"])
      oPort=portdict["oPort"];
}

string ExpansionValve::resultstring()
{
    string result;
    result = "\n" + name;
    result += "\n" + Port::title;
    result += "\n" + iPort->resultstring();
    result += "\n" + oPort->resultstring();
    return result;
}